src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/poptimizer/simple/CardinalityBasedGreedyJoinPlanOptimizerImpl.java

```diff
@@ -0,0 +1,291 @@
+
+            for ( int i = 1; i < subPlansWithStatistics.size(); i++ ) {
+                PhysicalPlanWithStatistics current = subPlansWithStatistics.get(i);
+                if ( current.getCandidate() < lowestCost ) {
+                    lowestCost = current.getCandidate();
+                    bestPlanWithCandidate = current;
+                }
+            }
+
+            return bestPlanWithCandidate;
+        }
+
+        /**
+         * The physical algorithm is determined based on the estimated number of requests to execute the join
+         * - SHJ: Card(T1)/PageSize + Card(T2)/PageSize
+         * - BJ: Card(T1)/PageSize + Card(T1)/blockSize
+         * For the comparison, only keeping the second element is enough.
+         */
+        protected PhysicalPlanWithStatistics decidePhysicalAlgorithm( final PhysicalPlanWithStatistics currentPlan, final PhysicalPlanWithStatistics nextPlan ) {
+            final double accNumSHJ = nextPlan.getNumOfAccess();
+
+            final List<FederationMember> fms = nextPlan.getFederationMembers();
+            double accNumBJ = 0;
+            for ( final FederationMember fm: fms ) {
+                double blockSize = setBlockSize(fm);
+                accNumBJ += currentPlan.getCandidate() / blockSize;
+            }
+
+            PhysicalPlan newPlan;
+            if ( accNumSHJ <= accNumBJ ) {
+                newPlan = PhysicalPlanFactory.createPlanWithJoin(currentPlan.plan, nextPlan.plan);
+            }
+            else {
+                newPlan = PhysicalPlanFactory.createPlanWithDefaultUnaryOpIfPossible(currentPlan.plan, nextPlan.plan);
+            }
+
+            final PhysicalPlanWithStatistics newPlanWithStatistics = new PhysicalPlanWithStatistics( newPlan );
+
+            final int joinCardinality = Math.min( currentPlan.getCandidate(), nextPlan.getCandidate() );
+            newPlanWithStatistics.addCandidate( joinCardinality );
+
+            return newPlanWithStatistics;
+        }
+
+        /**
+         * The number of requests depends on the page size of response.
+         */
+        protected int accessNumForReq( final int cardinality, final FederationMember fm ) {
+            if ( fm instanceof SPARQLEndpoint){
+                return 1;
+            }
+            else if ( (fm instanceof TPFServer) || (fm instanceof BRTPFServer) ){
+                final double pageSize = 100.0;
+                return (int) Math.ceil( cardinality/pageSize );
+            }
+            else
+                throw new IllegalArgumentException("Unsupported type of federation member: " + fm.getClass().getName() );
+        }
+
+        /**
+         * The block size (number of bindings can be attached) depends on the type of interface
+         */
+        protected double setBlockSize( final FederationMember fm ) {
+            if ( (fm instanceof SPARQLEndpoint) || (fm instanceof BRTPFServer) ){
+                return 30;
```

Comment:
Also here, instead of hard-coding the constant directly in the code, use the corresponding `static` member variable that we already have somewhere else.